cmd/cli/serve/util.go

```diff
@@ -218,3 +233,65 @@ func getNetworkConfig(nodeID string) (node.NetworkConfig, error) {
 		ClusterPeers:             networkCfg.Cluster.Peers,
 	}, nil
 }
+
+func getExecutionStore(ctx context.Context, storeCfg types.JobStoreConfig) (store.ExecutionStore, error) {
+	switch storeCfg.Type {
+	case types.BoltDB:
+		return boltdb.NewStore(ctx, storeCfg.Path)
+	default:
+		return nil, fmt.Errorf("unknown JobStore type: %s", storeCfg.Type)
+	}
+}
+
+func getJobStore(ctx context.Context, storeCfg types.JobStoreConfig) (jobstore.Store, error) {
+	switch storeCfg.Type {
+	case types.BoltDB:
+		log.Ctx(ctx).Debug().Str("Path", storeCfg.Path).Msg("creating boltdb backed jobstore")
+		return boltjobstore.NewBoltJobStore(storeCfg.Path)
+	default:
+		return nil, fmt.Errorf("unknown JobStore type: %s", storeCfg.Type)
+	}
+}
+
+func getNodeID(ctx context.Context) (string, error) {
```

Comment:
Once a nodeID is generated, it will be persisted in `config.yaml` and will be maintained during restarts. If the user deletes the repo, or modifies the config file manually, then this will look like a new node joining the network, which is similar to if the user deletes the libp2p key today.

Is that still going to be a problem?

Reply:
> Once a nodeID is generated, it will be persisted in config.yaml and will be maintained during restarts.

Where is this happening? I am missing where we update the config when the nodeID is set. 

Also, since node_name can be passed via the CLI as a flag or via an envvar wouldn't this mean if a user changed their node name with a flag I'd get a different auth key? I'd find that behavior unexpected.

Reply:
This is where it is happening https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/cmd/cli/serve/serve.go#L280

Yes, if the user run `bacalhau serve` with a different node name, then that is going to be a different node joining the network from the point of view of the requester. The mapping of executions to nodes is based on the node name/id, and will required different auth. Otherwise nodes can impersonate others by simply joining with their node names.

Though I am not sure how auth works today and how it is related to node names. Aren't we joining using a single nats token?

Reply:
> Once a nodeID is generated, it will be persisted in config.yaml and will be maintained during restarts.

That should meet the requirements of node-to-node auth, as long as we're happy that users changing the node name is equivalent to them creating a new node.

Tbh I think that is a little unexpected. What is the reason for allowing users to specifically set an ID? If it's to identify the nodes to humans, I would say that is conflating the node ID with its primary purpose, which is to identify nodes uniquely to the cluster. If we want to add human-readable identifiers, it would be more usual to do that in separate metadata.

Reply:
> Though I am not sure how auth works today and how it is related to node names. Aren't we joining using a single nats token?

The requester node will auto-generate a token that is expects to be used if one is not explicitly configured, and this token is generated by signing the node ID (a value unique to the node) using the node's private key (which only the node can do) – so that no other client can guess the token, the token is persistent, and there is little chance of a collision. As above, a user set node ID continues to allow this to work – changing the node ID (of the requester) would change the NATS auth token. If users want to migrate node IDs, they can just set an auth token explicitly (even the existing one).

Reply:
Prior to this PR:
The node ID was derived from a private key: https://github.com/bacalhau-project/bacalhau/blob/main/cmd/cli/serve/serve.go#L386
Requester nodes generate compute API auth secrets (token) by signing their node ID with their private key: https://github.com/bacalhau-project/bacalhau/blob/main/cmd/cli/serve/util.go#L186.

The above means that compute API auth secrets will remain the same so long as the requester node uses the same private key. The private key of a node is used to derive its identity (ID). 

------------------------------

In this PR (and the one merging into it):
The NodeID (which we now also refer to as the NodeName) can be provided by the user via a config flag, environment variable, or name provider. This means the node ID is easily subject to change. https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/cmd/cli/serve/util.go#L257

The Requester nodes still generate the API auth secrets by signing their node ID with their private key: https://github.com/bacalhau-project/bacalhau/blob/426-migrate-nodeID-boltdb-path/cmd/cli/serve/util.go#L201

The result of this change means that even if the private key for a node is unchanged, the compute API auth token can change if a different flag is used, different config file is used, env var is set, provider is altered, VM IP changes, etc. i.e. the compute API auth secrete is no longer derived from a nodes identity. 

-----------------------

This change as coupled the identity of a node with its name and will result in confusing behavior around auth. I would like to see these things de-coupled before merging.

Reply:
Lets take few steps back. 


> Tbh I think that is a little unexpected. What is the reason for allowing users to specifically set an ID? If it's to identify the nodes to humans, I would say that is conflating the node ID with its primary purpose, which is to identify nodes uniquely to the cluster. If we want to add human-readable identifiers, it would be more usual to do that in separate metadata.

Allowing operators to define their own human readable names is a nice feature to help them debug and manage their clusters. I understand we can add as many metadata to the node, but the node name is what operators will see in the logs, metrics, who is running an execution and commands such as `bacalhau node get <name/id>`. This is the case with Kubernetes btw where they allow operators to override the node name from using the `hostname` to a user provided value using [`--hostname-override`](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#:~:text=%2D%2Dhostname%2Doverride%20string). 
Though I am happy to remove the cli flag to make it not easy to change the node name to avoid mistakes where operators will have to set it in the config file.


> The requester node will auto-generate a token that is expects to be used if one is not explicitly configured, and this token is generated by signing the node ID (a value unique to the node) using the node's private key (which only the node can do) – so that no other client can guess the token, the token is persistent, and there is little chance of a collision. As above, a persistent user set or random node ID continues to allow this to work – changing the node ID (of the requester) would change the NATS auth token. If users want to migrate node IDs but still allow existing nodes to connect using their previous token, they can just set an auth token explicitly (even the existing one).

I am missing the value of the requester node using its ID to generate the token. How is this different than the requester node generating a random token at startup, persisting it, and then reuse it after restart regardless if the requester node name changes or not? As I've mentioned in [Node ACL](https://github.com/bacalhau-project/expanso-planning/issues/463), we need to decouple how we identify our nodes (ie. names), from how they do auth to enable different auth methods. 

I understand we are creating secure networks by default where we generate a token if one is missing instead of making the network open by default, but that would only work for testing purposes as this won't work when you have multiple requesters. In that environment, users would have to manually define their auth by either generating a token outside of the requesters if we continue to only support tokens, or just one of the other auth methods provided by NATS. Actually auto creation of tokens is even complicating creating clusters for testing purposes as we are now back to the point where you have to create the requester node first, inspect its output, and then create the compute nodes, which is similar to when we relied on the libp2p peerID for bootstrapping.

Lastly, we should persist the AuthSecret after it was generated and avoid regenerating it on the fly each time. If we decide to change our implementation, such as use a different hash instead of sha256, then all existing networks will break. This is a similar problem I faced with this PR where changing the default path of jobstore will lost state of all existing networks, and I had to implement a migration that persists the existing paths in the config file.

Btw, I do believe changing the node name, should change its identify and auth, but that should be handled in NATS where the node name would map to a NATS user, and each user would have limited permission to the subjects it can subscribe and publish to.

> This change has coupled the identity of a node with its name and will result in confusing behavior around auth. I would like to see these things de-coupled before merging.

I agree with the problem, but not with the solution. Answered above that we shouldn't generate the token based on the nodeID. In fact I don't think we should be auto generating tokens at all :) 

Btw, this was discussed in last week's [engineering standup](https://www.notion.so/expanso/Engineering-Standup-ed22dd09186f4d6bbe5055d5c8f075c8?pvs=4#75c47e59335f4fe8b5f8f5a0bbeeaf8b), but I understand these issues may not be obvious from the beginning.